libzim/libzim.pyx

```diff
@@ -0,0 +1,286 @@
+cimport libzim
+cimport cpython.ref as cpy_ref
+from cython.operator import dereference
+
+from libc.stdint cimport uint32_t, uint64_t
+from libcpp.string cimport string
+from libcpp cimport bool
+from libcpp.memory cimport shared_ptr, make_shared
+
+import datetime
+
+from collections import defaultdict
+
+#########################
+#       ZimBlob         #
+#########################
+
+cdef class ZimBlob:
+    cdef Blob* c_blob
+
+    def __init__(self, content):
+
+        if isinstance(content, str):
+            ref_content = content.encode('UTF-8')
+        else:
+            ref_content = content
+
+        self.c_blob = new Blob(<char *> ref_content, len(ref_content))
+
+    def __dealloc__(self):
+        if self.c_blob != NULL:
+            del self.c_blob
+
+
+#########################
+#       ZimArticle      #
+#########################
+
+cdef class ZimArticle:
+    cdef ZimArticleWrapper* c_article
+
+    def __init__(self):
+        self.c_article = new ZimArticleWrapper(<cpy_ref.PyObject*>self)
+    
+    def get_url(self):
+        raise NotImplementedError
+
+    def get_title(self):
+        raise NotImplementedError
+    
+    def is_redirect(self):
+        raise NotImplementedError
+
+    def get_mime_type(self):
+        raise NotImplementedError
+
+    def get_filename(self):
+        raise NotImplementedError
+
+    def should_compress(self):
+        raise NotImplementedError
+
+    def should_index(self):
+        raise NotImplementedError
+
+    def redirect_url(self):
+        raise NotImplementedError
+
+    def get_data(self):
+        raise NotImplementedError
+
+    @property
+    def mimetype(self):
+        return self.c_article.getMimeType().decode('UTF-8')
+    
+    @property
+    def content(self):
+        blob = self.c_article.getData()
+        content =  blob.data()[:blob.size()]
+        try:
+            return content.decode('UTF-8')
+        except UnicodeDecodeError:
+            return content
+
+    # This changes with implementation
+    @property
+    def can_write(self):
+        raise NotImplementedError
```

Comment:
This should be deleted. Probably a relic of the previous implementation.

Reply:
I am using can_write to safeguard writing incomplete articles in the creator. 

Each type of article has different demands as I understand. For instance, redirect can have empty content but needs a redirect url. If someone inherits from this and doesn't implement can_write it will blow before getting to add_article that can cause issues once inside libzim.

Reply:
Hum. Redirect articles must return true for `isRedirect`. If this is the case, libzim will only use the `getRedirectUrl` and not `getData`.

I'm not sure of what you want to protect with `can_write`. Is `can_write` is equivalent to `isRedirect` ? Or you want to protect of something else.

Reply:
I want to safeguard ZimCreator.add_article. From my point of view it's not 100% equivalent. 
For instance I was requiring a normal html article to have a Url and content. Redirect articles can required to have redirect url but can have empty content.

Reply:
Yes, but I don't understand when `can_write` would be false or true. The user code will always return true no ?
Having a `can_write` doesn't protect you from articles badly implemented.